### PR TITLE
Connect with firebase authentication

### LINKED ISSUE #1: POC Authentication with firebase admin
## Tasks
- [ ] Guide on connecting the front-end with Firebase API on the back-end
- [ ] Explanation of the required endpoint for Firebase integration
- [ ] Instructions on storing user information in our database, noting that Firebase only stores the `userId`

### PR DESCRIPTION
**Pull request by rmuti56**
**Topic**
Connect with firebase authentication

**Detail**

- Install firebase admin
- Create a new user via firebase auth

**Result**
Screenshot 2566-07-07 at 21 02 25
Screenshot 2566-07-07 at 21 04 28
Screenshot 2566-07-07 at 21 05 21

### CODE REVIEW

**Comment by rmuti56** [comment]
ref: #1

File: src/common/enums/role.enum.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export enum Role {
+  CUSTOMER = 'CUSTOMER',
```
Comment (thasup) [inline src/common/enums/role.enum.ts]: I prefer `USER` rather than `CUSTOMER`

File: src/migrations/1687092740504-create-user-table.ts [hunk 2]
```diff
@@ -17,19 +22,77 @@ export class CreateUserTable1687092740504 implements MigrationInterface {
...
+        referencedTableName: 'user',
+        referencedColumnNames: ['id'],
+        onDelete: 'CASCADE',
+      }),
+    );
   }
 
   public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.dropForeignKey('role', 'userId');
+    await queryRunner.dropTable('roles');
```
Comment (thasup) [inline src/migrations/1687092740504-create-user-table.ts]: don't any table name `roles`
not sure if it a typo or not?

File: src/modules/auth/auth.module.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { Module } from '@nestjs/common';
+import { AuthService } from './services/auth.service';
+import { AuthController } from './controllers/auth.controller';
+import { CoreModule } from 'src/core/core.module';
+import { UserModule } from '../user/user.module';
```
Comment (thasup) [inline src/modules/auth/auth.module.ts]: why sometimes use absolute import, and some use relative import?

File: src/modules/auth/services/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { BadRequestException, HttpStatus, Injectable } from '@nestjs/common';
+import { FirebaseAdminService } from 'src/core/firebase-admin/firebase-admin.service';
+import { CreateFirebaseUserDto } from '../dtos/create-firebase-user.dto';
```
Comment (thasup) [inline src/modules/auth/services/auth.service.ts]: should we using absolute import the same as the others?

File: src/modules/auth/services/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  async createUser(createFirebaseUserDto: CreateFirebaseUserDto) {
+    // validate exiting email
+    let user;
+    try {
+      user = await this.firebaseAdminService
+        .getAuth()
+        .getUserByEmail(createFirebaseUserDto.email);
+    } catch {
+      // do nothing
```
Comment (thasup) [inline src/modules/auth/services/auth.service.ts]: it's good to have `:TODO` word to indicate what was left for editing later.
easy to search, you know? 😘

File: src/modules/auth/services/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  constructor(
+    private readonly firebaseAdminService: FirebaseAdminService,
+    private readonly userService: UserService,
+  ) {}
+
+  async createUser(createFirebaseUserDto: CreateFirebaseUserDto) {
+    // validate exiting email
+    let user;
+    try {
+      user = await this.firebaseAdminService
```
Comment (thasup) [inline src/modules/auth/services/auth.service.ts]: if you don't use `user` elsewhere except in the if-condition. 
why don't you write it inside like this:

```
try {
  const user = await this.firebaseAdminService
      .getAuth()
      .getUserByEmail(createFirebaseUserDto.email);
  
  if (user) {
    throw new BadRequestException({
      code: ErrorCode.EMAIL_ALREADY_EXIST,
      statusCode: HttpStatus.BAD_REQUEST,
    });
  }
} catch  {
  // TODO: do nothing
}
```

File: src/modules/auth/services/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    if (user) {
+      throw new BadRequestException({
+        code: ErrorCode.EMAIL_ALREADY_EXIST,
+        statusCode: HttpStatus.BAD_REQUEST,
+      });
+    }
+
+    const newFirebaseUser = await this.firebaseAdminService
```
Comment (thasup) [inline src/modules/auth/services/auth.service.ts:29]: all async operations should wrap inside the try-catch block to handle errors in the future right?

who don't we wrap it all like these:

```
try {
  // check if user exists

  // create new firebase user

  // create new user from firebase email and uid

  // return new user
} catch {
  // TODO: do nothing
}
```

File: src/modules/user/dtos/create-user.dto.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export class CreateUserDto {
+  uid: string;
+
```
Comment (thasup) [inline src/modules/user/dtos/create-user.dto.ts]: remove unnecessary line

File: src/modules/user/services/user.service.ts [hunk 9]
```diff
@@ -1,15 +1,27 @@
...
-      name: 'test',
+  async createUser(createUserDto: CreateUserDto) {
+    const newUser = this.userRepository.create(createUserDto);
+
+    const user = await this.userRepository.save(newUser);
+    const newRole = this.roleRepository.create({
+      code: RoleEnum.CUSTOMER,
+      user: user,
     });
+    const role = await this.roleRepository.save(newRole);
```
Comment (thasup) [inline src/modules/user/services/user.service.ts:22]:
```
try {
  // save a new user

  // create a new role 

  // save a new role
} catch {
  // TODO: do nothing
}
```

**Review by thasup — CHANGES_REQUESTED** [review]
pls review my comments

**Review by thasup — APPROVED** [review]
LGTM

**Pull request merged**